### DISCUSSION TITLE
0039 XLS-39d: Clawback

### DISCUSSION THREAD

**Discussion by nbougalis** [XRPLF/XRPL-Standards#94] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by Mwni** [comment]
1. What is the reason for the auth and freeze modification flags? Can't this be done with a regular `TrustSet`?
2. Why should there be the `Reference` field when this could be put into a memo instead?

**Reply by nbougalis** [reply]
1. Sometimes it may be convenient to perform a group of operations _atomically_. In the absence of a "Multi-Transaction Transactor" (I have ideas on that!) the easiest way is to add support for some common-sense operations that are likely to be "bundled" together. By way of example, an issuer may be legally compelled to claw back a particular amount, but may also want to freeze the trust line out of an abundance of caution. Or, after initially freezing a trust line and investigating, the issuer decides to claw back some amount, and release the freeze.
2. Good question. If all you're looking to store is a small unique identifier, then a 256-bit value (or, really, a 128-bit [UUID](https://www.rfc-editor.org/rfc/rfc4122) value) is better because it is encoded more compactly than a memo is. But I'm not opposed to changing `Reference` to be a 128-bit value or eliminating altogether.

**Reply by Mwni** [reply]
Agreed on (1). But it's better to name these flags identical to the ones known from `TrustSet`. Namely `tfClearAuth`, `tfSetFreeze` and `tfClearFreeze`, to signal that these are identical operations. On (2), given how flexible the memo field is, the `Reference` field appears redundant.

**Reply by intelliot** [reply]
Agree with @Mwni : the `Reference` field should be removed from this spec, because it is unnecessary and redundant. Instead, there should be a Memo standard where the same reference data can be captured in a memo in a standardized way. The Reference Memo standard can be part of XLS-39, or it could be a separate spec - either way is fine.

**Reply by shawnxie999** [reply]
@intelliot could you elaborate on what you mean by capturing memo in a standardized way? What kind of change needs to be done with the existing Memo field?

**Reply by intelliot** [reply]
A spec needs to specify the format of the data that goes in the Memos. The Memos field is general and allows any arbitrary data to be stored. If that's fine, then you can leave it alone. But I think it is more likely that you'll want to specify a standard so that users and automated systems know how to interpret the data.

https://xrpl.org/transaction-common-fields.html#memos-field

For example, if you want to say that the `MemoType` must be `Reference`, then put that in a spec. Similarly, should the `MemoData` be a 256-bit value, a 128-bit UUID value, or something else? Also, how should it be encoded?

**Comment by dangell7** [comment]
I like this. I'm curious how this would work for the more advanced IOU functionality? Things like Escrow, PayChan and/or AMM (assuming they are passed).

**Reply by mDuo13** [reply]
I don't think it would have any interaction with AMM. As for escrow and payment channels with tokens, that's a good question. Should the issuer be able to clawback into the locked amount of the balance? Sounds complicated.

**Reply by shawnxie999** [reply]
AFAIK, escrow and payment channels are only for XRP.

**Reply by ckniffen** [reply]
@shawnxie999 for now but there is a PR for an amendment to add support

**Reply by shawnxie999** [reply]
@ckniffen do you have a link to the proposal or PR?

**Reply by shawnxie999** [reply]
nevermind i found it

**Reply by shawnxie999** [reply]
Just briefly read through XLS-34, I don't think it will cause any conflict with the Clawback functionality. However, after XLS-34 is live, Clawback won't be able to to claw back the funds that are locked in Escrow or Payment Channel. And if clawing back the funds that are locked is something that we want, then we would need to re-assess this option in more depth and have the feature update be part of another amendment after XLLS-34 goes live

**Comment by WietseWind** [comment]
I think the selected text below should not be `NFTokenMint` 😇 ?

<img width="794" alt="image" src="https://user-images.githubusercontent.com/4756161/222958254-238ece53-4de4-4bd1-9ddf-b8203cf7882f.png">

**Reply by nbougalis** [reply]
Oops, yes.

**Reply by ledhed2222** [reply]
LOL @nbougalis stockholm syndrome?

**Comment by WietseWind** [comment]
I'd like to propose an inverted flag: instead of **disabling** Clawback ability as an issuer with `lsfNoClawback`, have issuers **enable** explicitly with `lsfAllowClawback`, which can be set once.

I think this is a better way to grandfather in existing issuances / issuer accounts. This way existing issuers can only **enable** it once. Blackholed accounts used to issue tokens instead of IOUs cannot enable. Otherwise blackholed accounts will forever show as being able to Clawback, which may worry users. While they can't (blackholed), showing the issuing account as being able to is confusing.

Opposed to showing the existing (blackholed) accounts *cannot* Clawback, and also: cannot enable (blacholed).

**Reply by a-noni-mousse** [reply]
This is good and very similar to the handling of «disable rippleing» flag. Yes.

**Reply by Mwni** [reply]
Adding to this: If an issuer previously disabled freezing, indicating that they resign their ability to touch user funds, with the introduction of this amendment, all over sudden they have the ability again. This would somewhat breach the contract of trust with the holders.

**Reply by nbougalis** [reply]
Thanks @WietseWind. While I agree that, at first, your proposal seems like a good idea, I think it comes with problems (as does my version, to be clear).

Allowing clawback by default gives the issuer _new_ powers, that the holders of the token may not wish for the issuer to have. This is _somehow_ mitigated by the fact that there's already counterparty risk and the issuer may say "yes, your on-ledger balance is X quatloos, but I am going to withhold Y quatloos for _reasons_."

Disallowing clawback by default avoids granting issuers _new_ powers, _but_ at the same time, introduces a new concern: should an issuer be able to _enable_ clawback support if he has existing trustlines with outstanding balances? This is actually more complex in practice: since you can't iterate _all_ and analyze all trustlines, the presence of a _single_ trustline (or even a single owned object!) may be enough to prevent the functionality.

One option might be to allow the amendment corresponding to this feature to activate in two stages: in "stage 1" issuers gain the ability the _set_ the clawback flag. In "stage 2", activated some time later, the "clawback" operation becomes available and issuers can only set the clawback flag iff they don't have any existing trustlines. That way, people have time to decide if they want to continue holding potentially affected asset.

To follow up on @Mwni's thought, what if, when the `lsfNoFreeze` flag is set, the `lsfNoClawback` flag becomes implicit. That is:

```
bool canClawback(std::uint32_t flags)
{
    return (flags & (lsfNoFreeze | lsfNoClawback) == 0);
}
```

As a secondary issue, we can add a new 32-bit field into a payment `Initial Flags` that _ONLY_ comes into play if the payment does the initial funding of an account and specifies the flags that one wants to set on the account being created.

**Reply by Mwni** [reply]
Do we even need two seperate flags for freezing and clawback? If clawback just becomes an alternative to freezing with the same settings at inception, nobody could feel betrayed by this.

**Reply by nbougalis** [reply]
I don't _definitively_ know that we do, but I have opinions. While I'm open to repurposing `lsfNoFreeze` so that disclaiming freezing also disclaims clawback. But I don't want to _"limit"_ people too much by bundling several things into one "knob" either and the fact is that freezing has very different implications: it's all-or-nothing and prevents a much wider range of activities than clawback, which is an "one-time" (I know it cna be done repeatedly) adjustment as opposed to the "ongoing" action that freeze is.

**Reply by WietseWind** [reply]
Alternatively the default state of the clawback could be based on the freeze setting. If issuer already set NoFreeze, they automatically get NoClawback.

**Reply by nbougalis** [reply]
> Alternatively the default state of the clawback could be based on the freeze setting. If issuer already set NoFreeze, they automatically get NoClawback.

The challenge with that is that we can't go adjusting all `AccountRoot` objects to set the right flag when the amendment activates. Which leaves us with the option I mentioned earlier:

>when the `lsfNoFreeze` flag is set, the `lsfNoClawback` flag becomes implicit. That is:
>
>```
>bool canClawback(std::uint32_t flags)
>{
>    return (flags & (lsfNoFreeze | lsfNoClawback) == 0);
>}

**Reply by mDuo13** [reply]
I think I'm on board with redefining the NoFreeze flag to encompass clawback operations. Even though they're slightly different powers, they serve the same goals (regulatory compliance) and it seems to me that most people who want to have or give up one feature would make the same choice with the other one: either you want a "bank like" token with both clawback and freeze capabilities, or you want a "cash like" token with neither.

**Reply by WietseWind** [reply]
Ah sorry @nbougalis - you are right, your `return (flags & (lsfNoFreeze | lsfNoClawback) == 0)` does the trick indeed.

I second @mDuo13.

**Reply by shawnxie999** [reply]
I think incorporating clawback into freeze flag gives more flexibility to the issuer. Freeze allows the issuer to toggle clawback on/off on the trustline level as opposed to a `lsfNoClawback` that is on the account level. The current proposal grants everyone on the ledger the ability to clawback by default, and I don't think it is a good design to expose such a feature that is made specific for regulatory compliance to be enabled by default. Also I think forcing the issuer to freeze the trustline before clawing back is a good safety measure since clawback is an action that should be used with caution. So I definitely think that using the freeze flag to enable/disable clawback is really good from a safety measure perspective as well as from a design perspective.

If an account has set `lsfNoFreeze` flag, they probably don't care about clawback either since they have no intention of enforcing regulatory compliance

**Comment by mDuo13** [comment]
The `Reference` field seems functionally _very_ similar to the `InvoiceID` field of Payments. I guess that's intentional? Unfortunately `InvoiceID` is not quite the right name for this scenario, but we could still reuse the `InvoiceID` field here. (There is precedent for doing this anyway with Escrows' `OfferSequence` field.)

The tantalizing thing is that the canonical/binary format does not have a particular attachment to the fields' names, so it should be possible to implement aliases for field names. It's totally plausible we could reuse the field ID for `InvoiceID` but change the name in JSON, with no effects on the p2p network and how it processes the existing field—it would be harder to do this while maintaining backwards compatibility with client apps, though—you'd have to make the mapping of field name↔ID sensitive to the context of what transaction type it's in. (I know we previously talked about doing something like this by renaming the `Amount` field of Payments to `DeliverMax`, and nothing came of it; that's even more difficult because `Amount` is used in several transaction types and contexts.)

I would even consider changing the field name to be `Reference` _and_ making it a common field of all transactions so we don't have to worry about the need for this coming up on more transactions. I bet the community could find more uses for an arbitrary `Reference` ID that any transaction can include: for example, maybe a Hook could trigger a specific event when receiving a transaction with a hash matching the ID, or you could use the Reference with PaymentChannelClaim to reference a receipt or something that exhaustively lists what the Claims have been redeemed for, or something else like that.

I guess in the end it only matters if we're concerned about running out of 256-bit "uncommon field" IDs. If we aren't, it's not necessary to reuse identifiers across transactions, but it does feel like this scenario is niche enough that might be worth the awkwardness of "InvoiceID" as a field name to conserve more IDs for the future. We're not that close to running out of 256-bit fields yet, but they are one of the types we have the most of (second only to UInt32's) so it might be a prudent bit of future-proofing.

**Reply by WietseWind** [reply]
I agree @ `Reference` being a much better name than InvoiceID. The field _is_ being used though, so it could be a pretty tricky breaking change for some. But one worth it, imo. With the relatively low amount of use it would probably even be easy to track down the parties using it & informing them.

**Comment by mDuo13** [comment]
Another understated detail of this proposal is that it adds a way to disable the Auth flag of a trust line. Currently there is no way to revoke authorization of a trust line. I'm still not sure why it was designed that way, but I'm fairly certain it was a choice. Maybe @joelkatz or someone can clarify why.

In the absence of a counterargument I'm fine with adding that ability to clawbacks, but for consistency I think it would also make sense to modify TrustSet to add a flag to remove auth from a trust line too. (If this is controversial, maybe that flag should also be locked away if have No Freeze turned on.)

**Reply by shawnxie999** [reply]
Clear auth flag more seems like something that should belong to the TrustSet than Clawback. Not quite sure why this flag is included in this spec at all

**Reply by shawnxie999** [reply]
my recommendation is to leave out the clear auth flag from this spec. If we think this is something that should be added, it should be a separate change by itself, where `TrustSet` is updated and then `Clawback` if applicable

**Comment by shawnxie999** [comment]
I have updated the updated version of this spec in this [pull request](https://github.com/XRPLF/XRPL-Standards/pull/104)

The major change from the original spec here is that `Clawback` is made completely to be an extension of freeze, such that a trustline can only be clawed back if and only if the trustline has been frozen. Because of this, the `lsfNoClawback` flag is removed from the `AccountRoot`. 

It's still open to whether we still want to keep this flag or not, but I think for the sake of good design, both clawback and freeze should be bundled together as they are related for compliance. In a real world use case, token holder's trustline will be frozen first, then the issuer has the option to claw back from them.
